windows.c

```diff
@@ -0,0 +1,128 @@
+#include <ras.h>
+#include <tchar.h>
+#include <stdio.h>
+#include "common.h"
+
+void reportWindowsError(const char* action) {
+  LPTSTR pErrMsg = NULL;
+  DWORD errCode = GetLastError();
+  FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER|
+      FORMAT_MESSAGE_FROM_SYSTEM|
+      FORMAT_MESSAGE_ARGUMENT_ARRAY,
+      NULL,
+      errCode,
+      LANG_NEUTRAL,
+      pErrMsg,
+      0,
+      NULL);
+  printf("Error %s: %lu %s\n", action, errCode, pErrMsg);
+}
+
+// Figure out which Dial-Up or VPN connection is active; in a normal LAN connection, this should
+// return NULL. NOTE: For some reason this method fails when compiled in Debug mode but works
+// every time in Release mode.
+LPTSTR FindActiveConnection() {
+  DWORD dwCb = sizeof(RASCONN);
+  DWORD dwErr = ERROR_SUCCESS;
+  DWORD dwRetries = 5;
+  DWORD dwConnections = 0;
+  RASCONN* lpRasConn = NULL;
+  RASCONNSTATUS rasconnstatus;
+  rasconnstatus.dwSize = sizeof(RASCONNSTATUS);
+
+  //
+  // Loop through in case the information from RAS changes between calls.
+  //
+  while (dwRetries--) {
+    // If the memory is allocated, free it.
+    if (NULL != lpRasConn) {
+      HeapFree(GetProcessHeap(), 0, lpRasConn);
+      lpRasConn = NULL;
+    }
+
+    // Allocate the size needed for the RAS structure.
+    lpRasConn = (RASCONN*)HeapAlloc(GetProcessHeap(), 0, dwCb);
+    if (NULL == lpRasConn) {
+      dwErr = ERROR_NOT_ENOUGH_MEMORY;
+      break;
+    }
+
+    // Set the structure size for version checking purposes.
+    lpRasConn->dwSize = sizeof(RASCONN);
+
+    // Call the RAS API then exit the loop if we are successful or an unknown
+    // error occurs.
+    dwErr = RasEnumConnections(lpRasConn, &dwCb, &dwConnections);
+    if (ERROR_INSUFFICIENT_BUFFER != dwErr) {
+      break;
+    }
+  }
+  //
+  // In the success case, print the names of the connections.
+  //
+  if (ERROR_SUCCESS == dwErr) {
+    DWORD i;
+    for (i = 0; i < dwConnections; i++) {
+      RasGetConnectStatus(lpRasConn[i].hrasconn, &rasconnstatus);
+      if (rasconnstatus.rasconnstate == RASCS_Connected){
+        return lpRasConn[i].szEntryName;
+      }
+
+    }
+  }
+  return NULL; // Couldn't find an active dial-up/VPN connection; return NULL
+}
+
+int togglePac(bool turnOn, const char* pacUrl)
+{
+  int ret = RET_NO_ERROR;
+
+  INTERNET_PER_CONN_OPTION_LIST options;
+  DWORD   dwBufferSize = sizeof(options);
+  options.dwSize = dwBufferSize;
+  options.pszConnection = FindActiveConnection();
+
+  options.dwOptionCount = 2;
+  options.pOptions = (INTERNET_PER_CONN_OPTION*)calloc(2, sizeof(INTERNET_PER_CONN_OPTION));
+  if(!options.pOptions) {
+    return NO_MEMORY;
+  }
+
+  options.pOptions[0].dwOption = INTERNET_PER_CONN_FLAGS;
+  options.pOptions[1].dwOption = INTERNET_PER_CONN_AUTOCONFIG_URL;
+  if (turnOn) {
+    options.pOptions[0].Value.dwValue = PROXY_TYPE_AUTO_PROXY_URL;
+    options.pOptions[1].Value.pszValue = (char*)pacUrl;
+  }
+  else {
+    options.pOptions[0].Value.dwValue = PROXY_TYPE_DIRECT;
```

Comment:
Any suggestion? Should we add option to get configuration, so Lantern can hold it somewhere and reapply when quit? But other tools may also change proxy setting during this period. It did occur to me when I had both Shadowsocks and Lantern opened.